### ISSUE TITLE
Creating a new workspace on monitor if using `m+1`

### ISSUE THREAD

**Issue by rxt30** [hyprwm/Hyprland#1473]
Hi,

i would like to suggest the feature, that if using a `relative workspace to monitor (m+1)` argument in a `workspace` or `movetoworkspace/movetoworkspacesilent` dispatcher, a new workspace should be created instead of doing nothing in the case of the `workspace` dispatcher.
As an alternative `monitor+next` and `monitor+previous` as a new workspace choice may also be an option, although they imo do not differ that much from `m+1` and `m-1`.

This could be beneficial to users in a multi-monitor setup, as they could just move back and forth between workspaces on a monitor instead of having to assign workspaces to outputs or remember the assigned workspaces.

TL;DR expected behaviour
If a monitor has only one workspace and `hyperctl dispatch workspace m+1` is called, a new empty workspace should be created

**Comment by haininhhoang94** [comment]
* Temporary solution
1. Create a workspace.py file
`#!/usr/bin/python
import subprocess
import time
import json


def current_focus_workspace():
    workspace_list = (
        subprocess.run("hyprctl -j monitors".split(), stdout=subprocess.PIPE)
        .stdout.decode("utf-8")
        .strip()
    )

    workspace_list = json.loads(workspace_list)
    for workspace in workspace_list:
        if workspace["focused"] == True:
            return [workspace["name"], workspace["activeWorkspace"]["id"]]
    return []


i = int(current_focus_workspace()[1]) + 1
j = current_focus_workspace()[0]
subprocess.run(f"hyprctl dispatch workspace {i}".split())
subprocess.run(f"hyprctl dispatch moveworkspacetomonitor {i} {j}".split())
`
2. Bind the key:

> bind = $HLMOD, D, exec, python ~/.config/hypr/scripts/scripts/workspaces.py

Let me know if you want something more!

**Comment by ConnorWay32** [comment]
Would prefer if something like this was made into its own dispatch command, since m+1 is useful for cycling through workspaces on a monitor, and this suggestion would create an empty workspace, making it less streamlined.

**Comment by CheesyPhoenix** [comment]
To expand on the answer by @haininhhoang94:

If you save this python script as ``next-workspace.py`` in  ``.config/hypr/scripts``:
```py
#!/usr/bin/python
import subprocess
import json
import sys


def current_focus_workspace():
    monitor_list = (
        subprocess.run("hyprctl -j monitors".split(), stdout=subprocess.PIPE)
        .stdout.decode("utf-8")
        .strip()
    )

    monitor_list = json.loads(monitor_list)
    for monitor in monitor_list:
        if monitor["focused"] == True:
            return [monitor["name"], monitor["activeWorkspace"]["id"]]
    return []


currWorkspace = int(current_focus_workspace()[1])
currMonitor = current_focus_workspace()[0]

workspacesList = json.loads(subprocess.run(
    "hyprctl -j workspaces".split(), stdout=subprocess.PIPE)
    .stdout.decode("utf-8")
    .strip()
)

workSpacesOfMonitor = list(filter(
    lambda workspace: workspace["monitor"] == currMonitor, workspacesList))

workSpacesOfMonitor.sort(
    key=(lambda workspace: int(workspace["id"])), reverse=True)

highestWorkspaceOnMonitor = workSpacesOfMonitor[0]

if highestWorkspaceOnMonitor["id"] > currWorkspace:
    subprocess.run("hyprctl dispatch workspace m+1".split())
else:
    workspacesList.sort(
        key=(lambda workspace: workspace["id"]), reverse=True)
    highestWorkspaceID = workspacesList[0]["id"]

    if "--with-window" in sys.argv:
        subprocess.run(
            f"hyprctl dispatch movetoworkspace {highestWorkspaceID + 1}".split())
    else:
        subprocess.run(
            f"hyprctl dispatch workspace {highestWorkspaceID + 1}".split())
    subprocess.run(
        f"hyprctl dispatch moveworkspacetomonitor {highestWorkspaceID + 1} {currMonitor}".split())
```

And then add these four binds to your ``hyprland.conf``:
```conf
bind = $mainMod, left, workspace, m-1
bind = $mainMod SHIFT, left, movetoworkspace, m-1
bind = $mainMod, right, exec, python ~/.config/hypr/scripts/next-workspace.py
bind = $mainMod SHIFT, right, exec, python ~/.config/hypr/scripts/next-workspace.py --with-window
```

You can then navigate between each monitor's workspaces, creating new ones when neccessary, by holding $mainMod and pressing arrow left or right. If you hold down shift you will also take your currently focused window with you.

**Comment by hvitoi** [comment]
+1 for this as a built-in feature.
This is the default navigating experience on gnome and other people coming from this DE might also be interested in it.
@CheesyPhoenix thanks for the script. One thing I would add here is also changing the behavior of `m-1` to do nothing when it's the first workspace (instead of moving it to the last workspace)

**Comment by CheesyPhoenix** [comment]
> +1 for this as a built-in feature.
> This is the default navigating experience on gnome and other people coming from this DE might also be interested in it.
> @CheesyPhoenix thanks for the script. One thing I would add here is also changing the behavior of `m-1` to do nothing when it's the first workspace (instead of moving it to the last workspace) 
> 

I have now created a small program for this in rust that also handles moving backwards. Its installable with cargo aswell as from the AUR. See: https://github.com/CheesyPhoenix/hyprland-relative-workspace

**Comment by mshzsh** [comment]
Inspired by Script @CheesyPhoenix 
with empty workspaces support.

```
#!/usr/bin/python
import subprocess
import json
import sys

cli = 'movetoworkspace' if "--move" in sys.argv else 'workspace'


def get_current_workspace():
    monitor_list = (
        subprocess.run("hyprctl -j monitors".split(), stdout=subprocess.PIPE, check=False)
        .stdout.decode("utf-8")
        .strip()
    )

    monitor_list = json.loads(monitor_list)
    for monitor in monitor_list:
        if monitor["focused"] == True:
            return [monitor["name"], monitor["activeWorkspace"]["id"]]
    return []


current_workspace = int(get_current_workspace()[1])
current_monitor = str(get_current_workspace()[0])


workspaces = json.loads(subprocess.run(
    "hyprctl -j workspaces".split(), stdout=subprocess.PIPE, check=False)
    .stdout.decode("utf-8")
    .strip()
)

current_monitor_workspaces = []
all_workspaces = []
for ws in workspaces:
    all_workspaces.append(int(ws['id']))
    if ws['monitor'] == current_monitor:
        current_monitor_workspaces.append(int(ws['id']))
current_monitor_workspaces.sort(reverse=False)
all_workspaces.sort(reverse=False)

cmw_with_empty = []
for i in range(1, all_workspaces[-1]+1):
    if i in current_monitor_workspaces or (i not in current_monitor_workspaces and i not in all_workspaces):
        cmw_with_empty.append(i)


if "--right" in sys.argv:
    if current_workspace == cmw_with_empty[-1]:
        subprocess.run(
            f"hyprctl dispatch {cli} {all_workspaces[-1]+1}".split())
    else:
        count = 0
        for cmw in cmw_with_empty:
            if cmw == current_workspace:
                subprocess.run(
                    f"hyprctl dispatch {cli} {cmw_with_empty[count+1]}".split())
                break
            count += 1

elif "--left" in sys.argv:
    if current_workspace != cmw_with_empty[0]:
        count = 0
        for cmw in cmw_with_empty:
            if cmw == current_workspace:
                subprocess.run(
                    f"hyprctl dispatch {cli} {cmw_with_empty[count-1]}".split())
                break
            count += 1

```

- bind=SUPER,LEFT,exec, python ~/.config/hypr/navigate_workspaces.py --left
- bind=SUPER,RIGHT,exec, python ~/.config/hypr/navigate_workspaces.py --right
- bind=SUPERCTRL,LEFT,exec, python ~/.config/hypr/navigate_workspaces.py --left --move
- bind=SUPERCTRL,RIGHT,exec, python ~/.config/hypr/navigate_workspaces.py --right --move

**Comment by hvitoi** [comment]
The [r+1/r-1](https://wiki.hyprland.org/Configuring/Dispatchers/#workspaces) workspace alias with empty workspaces support has been recently pushed.

**Comment by vaxerski** [comment]
yeah, closing because of the above
